### ISSUE TITLE
[HttpKernel] `#[MapRequestPayload]` is "handled" before `#[IsGranted]`

### ISSUE THREAD

**Issue by artyuum** [symfony/symfony#50120]
### Symfony version(s) affected

6.3

### Description

After playing a bit with [this new feature](https://symfony.com/blog/new-in-symfony-6-3-mapping-request-data-to-typed-objects), I noticed that the payload is getting mapped/validated **before** ensuring that the user is actually allowed to access the resource.

This seems odd to me and I'd like to know your opinion about that. 

### How to reproduce

1. create a route and its payload.
2. use the `#[MapRequestPayload]` attribute on the payload argument (in the controller)
3. secure the route using the `#[IsGranted]` attribute
4. access the route by posting some data (voluntarily malformed data) but without being logged-in
5. see the error message related to the malformed data (and not because the user is not logged-in)

### Possible Solution

Few years ago, I created a bundle to achieve the same thing that this feature is offering, and I encountered the same problem. 
I played around with the "priority" option but without luck. So my only solution was to create a listener on the [ControllerArgumentsEvent](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/HttpKernel/Event/ControllerArgumentsEvent.php) instead of using an ArgumentResolver. That way, the `#[IsGranted]` attribute would always be fired before my listener. No clue if I did things right at the time, but that's what worked for me.

**Comment by nicolas-grekas** [comment]
Nice catch.

This happens because arguments are resolved before `kernel.controller_arguments` is dispatched of course.
This relates to exceptions in argument value resolvers: in #49138, we couldn't generate responses directly for them. Instead, we had to throw exceptions and improve our ProblemNormalizer in #49944 to return proper 4xx.

Maybe we need an architectural change here. One idea could be to let the resolver do only the first part of the resolution, by "reserving" the argument for a corresponding listener we'd add on `kernel.controller_arguments`. My idea (without looking at the code) would be to resolve the argument **to the instance of the attribute**, and have the listener turn that into proper objects, using serializer+validator. There, we could also improve error reporting, by generating the 4xx inline.

/cc @Koc WDYT?

Up to give it a try?

**Comment by nicolas-grekas** [comment]
@artyuum can you please try this PR? #50125

**Comment by artyuum** [comment]
@nicolas-grekas I cloned your repo, checked out your branch and used the `link` script to copy that specific version of Symfony into a working Sf project and I'm having the following error (even after nuking the cache):
> Attempted to call an undefined method named "onKernelControllerArguments" of class "Symfony\Component\HttpKernel\Controller\ArgumentResolver\TraceableValueResolver".

<details><summary>Stacktrace</summary>

```text
Symfony\Component\ErrorHandler\Error\UndefinedMethodError:
Attempted to call an undefined method named "onKernelControllerArguments" of class "Symfony\Component\HttpKernel\Controller\ArgumentResolver\TraceableValueResolver".

  at /var/www/html/test/vendor/symfony/event-dispatcher/Debug/WrappedListener.php:116
  at Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke()
     (/var/www/html/test/vendor/symfony/event-dispatcher/EventDispatcher.php:220)
  at Symfony\Component\EventDispatcher\EventDispatcher->callListeners()
     (/var/www/html/test/vendor/symfony/event-dispatcher/EventDispatcher.php:56)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch()
     (/var/www/html/test/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php:139)
  at Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch()
     (/var/www/html/test/vendor/symfony/http-kernel/HttpKernel.php:161)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw()
     (/var/www/html/test/vendor/symfony/http-kernel/HttpKernel.php:74)
  at Symfony\Component\HttpKernel\HttpKernel->handle()
     (/var/www/html/test/vendor/symfony/http-kernel/EventListener/ErrorListener.php:108)
  at Symfony\Component\HttpKernel\EventListener\ErrorListener->onKernelException()
     (/var/www/html/test/vendor/symfony/event-dispatcher/Debug/WrappedListener.php:116)
  at Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke()
     (/var/www/html/test/vendor/symfony/event-dispatcher/EventDispatcher.php:220)
  at Symfony\Component\EventDispatcher\EventDispatcher->callListeners()
     (/var/www/html/test/vendor/symfony/event-dispatcher/EventDispatcher.php:56)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch()
     (/var/www/html/test/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php:139)
  at Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch()
     (/var/www/html/test/vendor/symfony/http-kernel/HttpKernel.php:224)
  at Symfony\Component\HttpKernel\HttpKernel->handleThrowable()
     (/var/www/html/test/vendor/symfony/http-kernel/HttpKernel.php:117)
  at Symfony\Component\HttpKernel\HttpKernel->terminateWithException()
     (/var/www/html/test/vendor/symfony/http-kernel/EventListener/DebugHandlersListener.php:74)
  at Symfony\Component\HttpKernel\EventListener\DebugHandlersListener::Symfony\Component\HttpKernel\EventListener\{closure}()
     (/var/www/html/test/vendor/symfony/error-handler/ErrorHandler.php:537)
  at Symfony\Component\ErrorHandler\ErrorHandler->handleException()
```

</details> 

Do you have the same error?

**Comment by artyuum** [comment]
@nicolas-grekas with this PR I can't reproduce the issue 👍🏻

**Comment by lyrixx** [comment]
It's not that simple IMHO. What if you need the incoming payload in the security layer?

for example, I added this feature in ApiPlatform https://github.com/api-platform/core/pull/4392

So I think the best option is to be able to defined the priority in the Attribute itself, and even better: Allow to set two IsGranted attributes

```php
#[IsGranted('ROLE_API_USER', priotity: 10)
#[IsGranted('BLOG_POST_WRITE', priotity: 0)
public function write(
    #MapRequestBody(priotity: 5)
    BlogPost $blogPost
) {
}
```

**Comment by ro0NL** [comment]
i tend to believe deserializing the user payload before authing is generally fine

what should not happen is leaking the validation errrors before authing, so maybe defer that?

**Comment by artyuum** [comment]
I have a different opinion about this. Users are free to create a custom constraint and some constraints can be resource intensive. My thought is that an unauthorized user can repeatedly try to access a route and thus trigger the validators for nothing, while they aren't supposed to have access to that route.

Doing the mapping/validation even before knowing if the user is allowed to access the route seems like a security issue to me.

**Comment by derrabus** [comment]
> I noticed that the payload is getting mapped/validated **before** ensuring that the user is actually allowed to access the resource.

For reference: The good old Doctrine entity param converters have worked that way and that was desired because we might need to load the entity before we can decide if the user is actually permitted to access it.

> Doing the mapping/validation even before knowing if the user is allowed to access the route seems like a security issue to me.

Simply deserializing or validating the payload should be fine as long as we don't persist anything, or am I missing something? Can you elaborate?

**Comment by artyuum** [comment]
> Simply deserializing or validating the payload should be fine as long as we don't persist anything

My argument is not about that but about the unnecessary execution of slow / resource-intensive constraints. "Native" constraints are pretty lightweight, but we must keep in mind that customized constraints can be created. Someone can send a malformed/invalid payload to a route, without even being authenticated and all of the constraints will be executed (the developer can leverage the `Sequentially` constraint to *reduce* the impact though)

Another example of bad intention that can be done is when calling a third-party API inside a validator to validate the data using an external service, someone can repeatedly trigger that constraint's validator (thus making unnecessary API calls) without even being authenticated. Depending on the API used, a rate limit between the third-party API and the Symfony application could happen and negatively affect legitimate requests.

**Comment by derrabus** [comment]
>  calling a third-party API inside a validator

Holy smokes, I keep underestimating the creativity of developers sometimes. 😓

**Comment by artyuum** [comment]
But this is a legitimate use case and there is nothing wrong with it Alexander 😅 Some developers are required to validate some data using an external service. One example can be [email verification](https://www.mailgun.com/products/inbox/email-verification) but there are many more use cases besides email verification.

**Comment by ro0NL** [comment]
what about doing validation in 2 steps? respectively pre- and post-auth.

we can introduce a new default validation group, so we have `Default` + `Authenticated` or so

**Comment by AurelienPillevesse** [comment]
I agree with artyuum, unnecessary treatments should not happen, in this case: the validation before the is_granted.

One solution might be to say: 
Isgranted attribute on function can not have a parameter (only "ROLE_ADMIN" for example), if you want to use it with a parameter like a validated payload : call the isgranted function inside your controller action but not with the PHP8 attribute.

What do you think about this solution ?

**Comment by derrabus** [comment]
> One solution might be to say: Isgranted attribute on function can not have a parameter

But it can already have a parameter. Your "solution" breaks apps that already work perfectly fine.

**Comment by ro0NL** [comment]
maybe the whole thing should happen lazy, so deserialize+validate on-demand during security, or finally before invoking the controller

**Comment by nicolas-grekas** [comment]
Please check https://github.com/symfony/symfony/pull/50125#pullrequestreview-1397289473
TL;DR I think we should fix this and think about ways to allow accessing payloads from IsGranted separately, for a future version.
